src/IO/SeekableStreamingReadBuffer.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <IO/ReadBufferFromFileBase.h>
+
+
+namespace DB
+{
+
+class SeekableStreamingReadBuffer : public ReadBufferFromFileBase
+{
+    std::unique_ptr<ReadBufferFromFileBase> nested;
+    UInt64 read_seek_threshold;
+
+public:
+    SeekableStreamingReadBuffer(std::unique_ptr<ReadBufferFromFileBase> nested_, UInt64 read_seek_threshold_)
+        : nested(std::move(nested_))
+        , read_seek_threshold(read_seek_threshold_)
+    {
+        swap(*nested);
+    }
+
+    std::string getFileName() const override { return nested->getFileName(); }
+
+    off_t getPosition() override
+    {
+        swap(*nested);
+        off_t position = nested->getPosition();
+        swap(*nested);
+        return position;
+    }
+
+    off_t seek(off_t off, int whence) override
+    {
+        off_t position = getPosition();
+
+        if (whence == SEEK_CUR)
+        {
+            off += position;
+            whence = SEEK_SET;
+        }
+
+        if (whence == SEEK_SET && off >= position && off < position + static_cast<off_t>(read_seek_threshold))
+        {
+            swap(*nested);
+            nested->ignore(off - position);
```

Comment:
Ok. But we already have this optimization in at least two places:

1. If seek position is already in buffer, we simply move the cursor.
```
    /// Position is unchanged.
    if (new_pos + (working_buffer.end() - pos) == pos_in_file)
        return new_pos;

    if (hasPendingData() && new_pos <= pos_in_file && new_pos >= pos_in_file - static_cast<off_t>(working_buffer.size()))
    {
        /// Position is still inside buffer.
        pos = working_buffer.begin() + (new_pos - (pos_in_file - working_buffer.size()));
        return new_pos;
    }
```

2. Settings to avoid small seeks:
```
    M(SettingUInt64, merge_tree_min_rows_for_seek, 0, "You can skip reading more than that number of rows at the price of one seek per file.", 0) \
    M(SettingUInt64, merge_tree_min_bytes_for_seek, 0, "You can skip reading more than that number of bytes at the price of one seek per file.", 0) \
```
(they are set to zero because they don't help in typical case)

Reply:
If you will set these settings, you may experience similar performance benefit.

Reply:
But that strategy shall depend on type and speed of disk!

Reply:
I wonder if `merge_tree_min_rows_for_seek` is useful at all. Because rows can be of different size, and `min_bytes` shall be calculated as `speed * seek time`.